File: backend/src/main/java/com/vira/dply/service/AuthService.java
package com.vira.dply.service;

import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import com.vira.dply.entity.UserEntity;
import com.vira.dply.enums.Role;
import com.vira.dply.repository.UserRepository;

import lombok.RequiredArgsConstructor;

@Service
@RequiredArgsConstructor
public class AuthService {

    private final UserRepository userRepository;
    private final PasswordEncoder passwordEncoder;
    private final JwtService jwtService;

    public String register(String email, String password) {
        if (userRepository.existsByEmail(email)) {
            throw new IllegalArgumentException("Email already registered");
        }

        UserEntity user = new UserEntity();
        user.setEmail(email);
        user.setPasswordHash(passwordEncoder.encode(password));
        user.setRole(Role.VIEWER); // rol inicial por defecto
        userRepository.save(user);

        return jwtService.generateToken(user.getId(), user.getEmail());
    }

    public String login(String email, String password) {
        UserEntity user = userRepository.findByEmail(email)
                .orElseThrow(() -> new IllegalArgumentException("Invalid credentials"));

        if (!passwordEncoder.matches(password, user.getPasswordHash())) {
            throw new IllegalArgumentException("Invalid credentials");
        }

        return jwtService.generateToken(user.getId(), user.getEmail());
    }
}

File: backend/src/main/kotlin/com/vira/dply/service/AuthService.kt
package com.vira.dply.service

import com.vira.dply.model.User
import com.vira.dply.model.UserRole
import com.vira.dply.repository.UserRepository
import org.springframework.security.authentication.AuthenticationManager
import org.springframework.security.authentication.UsernamePasswordAuthenticationToken
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.crypto.password.PasswordEncoder
import org.springframework.stereotype.Service


@Service
class AuthService(
    val userRepository: UserRepository,
    val passwordEncoder: PasswordEncoder,
    val authenticationManager: AuthenticationManager,
    val jwtService: JwtService,
) {

    fun register(user: User): User {
        user.encodedPassword = passwordEncoder.encode(user.password).toString()
        user.roles = listOf(UserRole.ROLE_USER)

        return userRepository.save(user)
    }

    fun login(email: String, password: String): String {
        authenticationManager.authenticate(
            UsernamePasswordAuthenticationToken(email, password)
        )

        val user: User = userRepository.findByEmail(email)

        return jwtService.generateToken(user)
    }
}